pkg/client/unversioned/discovery_client.go

```diff
@@ -0,0 +1,175 @@
+// preferred version.
+func (d *DiscoveryClient) ServerGroups() (apiGroupList *unversioned.APIGroupList, err error) {
+	// Get the groupVersions exposed at /api
+	d.baseURL.Path = "/api"
+	resp, err := d.get(d.baseURL.String())
+	if err != nil {
+		return nil, err
+	}
+	var v unversioned.APIVersions
+	defer resp.Body.Close()
+	err = json.NewDecoder(resp.Body).Decode(&v)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error: %v", err)
+	}
+	apiGroup := apiVersionsToAPIGroup(&v)
+
+	// Get the groupVersions exposed at /apis
+	d.baseURL.Path = "/apis"
+	resp2, err := d.get(d.baseURL.String())
+	if err != nil {
+		return nil, err
+	}
+	defer resp2.Body.Close()
+	apiGroupList = &unversioned.APIGroupList{}
+	if err = json.NewDecoder(resp2.Body).Decode(&apiGroupList); err != nil {
+		return nil, fmt.Errorf("unexpected error: %v", err)
+	}
+
+	// append the group retrieved from /api to the list
+	apiGroupList.Groups = append(apiGroupList.Groups, apiGroup)
+	return apiGroupList, nil
+}
+
+// APIResourcesForGroup returns the supported resources for a group and version.
+func (d *DiscoveryClient) ServerResourcesForGroupVersion(groupVersion string) (resources *unversioned.APIResourceList, err error) {
+	if groupVersion == "v1" {
+		d.baseURL.Path = "/api/" + groupVersion
+	} else {
+		d.baseURL.Path = "/apis/" + groupVersion
+	}
+	resp, err := d.get(d.baseURL.String())
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	resources = &unversioned.APIResourceList{}
+	if err = json.NewDecoder(resp.Body).Decode(resources); err != nil {
+		return nil, fmt.Errorf("unexpected error: %v", err)
+	}
+	return resources, nil
+}
+
+// APIResources returns the supported resources for all groups and versions.
+func (d *DiscoveryClient) ServerResources() (map[string]*unversioned.APIResourceList, error) {
+	apiGroups, err := d.ServerGroups()
+	if err != nil {
+		return nil, err
+	}
+	groupVersions := extractGroupVersions(apiGroups)
+	result := map[string]*unversioned.APIResourceList{}
+	for _, groupVersion := range groupVersions {
+		resources, err := d.ServerResourcesForGroupVersion(groupVersion)
+		if err != nil {
+			return nil, err
+		}
+		result[groupVersion] = resources
+	}
+	return result, nil
+}
+
+func setDiscoveryDefaults(config *Config) error {
+	config.Prefix = ""
+	config.Version = ""
+	return nil
+}
+
+// NewDiscoveryClient creates a new DiscoveryClient for the given config. This client
+// can be used to discover supported resources in the API server.
+func NewDiscoveryClient(c *Config) (*DiscoveryClient, error) {
+	config := *c
+	if err := setDiscoveryDefaults(&config); err != nil {
+		return nil, err
+	}
+	transport, err := TransportFor(c)
+	if err != nil {
+		return nil, err
+	}
+	client := &http.Client{Transport: transport}
+	baseURL, err := defaultServerUrlFor(c)
+	return &DiscoveryClient{client, *baseURL}, nil
+}
+
+// NewDiscoveryClientForRESTCLient constructs a DiscoveryClient based on a RESTClient.
```

Comment:
It's not being used currently. I added it according to lavalamp's client package proposal (https://github.com/kubernetes/kubernetes/pull/15730/files#diff-db2b1446b41ce62918b3fc288d73dad0R81)